theano/tensor/opt.py

```diff
@@ -3987,11 +3988,32 @@ def local_useless_reshape(node):
     single dimension.
 
     """
-    if isinstance(node.op, T.Reshape):
-        if (node.inputs[0].ndim == 1 and node.outputs[0].ndim == 1 and
-                node.inputs[0].broadcastable ==
-                node.outputs[0].broadcastable):
-            return [node.inputs[0]]
+    op = node.op
+    if not isinstance(op, Reshape):
+        return False
+
+    input = node.inputs[0]
+    output = node.outputs[0]
+    output_shape = node.inputs[1]
+
+    if (input.ndim == 1 and output.ndim == 1 and
+            input.broadcastable == output.broadcastable):
+        return [input]
+
+    dimshuffle_new_order = []
+    new_output_shape = []
+    index = 0  # index over the output of the new reshape
+    for i in xrange(output.ndim):
+        dim = extract_constant(output_shape[i], only_process_constants=False)
+        if dim == 1:
+            dimshuffle_new_order.append('x')
+        else:
+            dimshuffle_new_order.append(index)
+            new_output_shape.append(dim)
+            index = index + 1
+    if index != output.ndim:
+        inner = op.__class__(len(new_output_shape))(input, new_output_shape)
```

Comment:
I would make the same questions here. Why do we want to make explicitly the DimShuffle when the Reshape isn't removed? When it is removed, I agree this is good, but when we keep it. I don't understand why we need to keep it.

Also, if we end up making the case where we want to have an dimshuffle and a reshape when we keep both, I would special case here the case where we remove the reshape, to not let another opt do the clean up. This will speed up optimization.